src/test/java/marquez/MarquezAppIntegrationTest.java

```diff
@@ -461,6 +461,99 @@ public void testApp_createJobAndMarkRunAsComplete() {
+    final SourceMeta sourceMeta =
+        SourceMeta.builder()
+            .type(STREAM_SOURCE_TYPE)
+            .connectionUrl(STREAM_CONNECTION_URL)
+            .description(STREAM_SOURCE_DESCRIPTION)
+            .build();
+
+    client.createSource(STREAM_SOURCE_NAME, sourceMeta);
+
+    // Create job
+    final JobMeta jobMeta =
+        JobMeta.builder()
+            .type(JOB_TYPE)
+            .inputs(ImmutableSet.of())
+            .outputs(ImmutableSet.of())
+            .location(JOB_LOCATION)
+            .context(JOB_CONTEXT)
+            .description(JOB_DESCRIPTION)
+            .build();
+
+    final Job job = client.createJob(NAMESPACE_NAME, jobName, jobMeta);
+
+    // Create a run
+    final RunMeta runMeta = RunMeta.builder().build();
+    final Run run = client.createRun(NAMESPACE_NAME, jobName, runMeta);
+    assertThat(run.getId()).isNotNull();
+    assertThat(run.getCreatedAt()).isAfter(EPOCH);
+    assertThat(run.getUpdatedAt()).isAfter(EPOCH);
+    assertThat(run.getNominalStartTime()).isEmpty();
+    assertThat(run.getNominalEndTime()).isEmpty();
+    assertThat(run.getState()).isEqualTo(RunState.NEW);
+    assertThat(run.getStartedAt()).isEmpty();
+    assertThat(run.getEndedAt()).isEmpty();
+    assertThat(run.getDurationMs()).isEmpty();
+    assertThat(run.getArgs()).isEmpty();
+
+    // Create datasets
+    final ImmutableSet<DatasetId> inputs = newDatasetIdsWith(NAMESPACE_NAME, 2);
+    for (final DatasetId input : inputs) {
+      final DbTableMeta dbTableMeta =
+          DbTableMeta.builder()
+              .physicalName(input.getName())
+              .sourceName(STREAM_SOURCE_NAME)
+              .description(newDescription())
+              .build();
+
+      client.createDataset(input.getNamespace(), input.getName(), dbTableMeta);
+    }
+
+    final ImmutableSet<DatasetId> outputs = newDatasetIdsWith(NAMESPACE_NAME, 2);
+    for (final DatasetId output : outputs) {
+      final DbTableMeta dbTableMeta =
+          DbTableMeta.builder()
+              .physicalName(output.getName())
+              .sourceName(STREAM_SOURCE_NAME)
+              .runId(run.getId()) // with run ID in output dataset
+              .description(newDescription())
+              .build();
+
+      client.createDataset(output.getNamespace(), output.getName(), dbTableMeta);
+    }
+
+    // Update job
+    final JobMeta jobUpdateMeta =
+        JobMeta.builder()
+            .type(JOB_TYPE)
+            .inputs(inputs)
+            .outputs(outputs)
+            .runId(runMeta.getId())
+            .location(JOB_LOCATION)
+            .context(JOB_CONTEXT)
+            .description(JOB_DESCRIPTION)
+            .build();
+    client.createJob(NAMESPACE_NAME, jobName, jobUpdateMeta);
+
+    // Assure datasets are associated
+    Job finalJob = client.getJob(NAMESPACE_NAME, jobName);
+    assertThat(finalJob.getInputs()).hasSize(2);
+
+    // todo: Assure new input dataset has been associated.
```

Comment:
I think you could pass a mock listener through the config.
Or just add a separate test where you mock the listener in the JobServiceTest